### PR TITLE
fix(validation): Sprint 1-2 Pass 1 emulator defects (D1-D14)

### LINKED ISSUE #95: [S1] iOS app crashes on launch: flutter_contacts 1.1.9+2 nil-unwraps AppDelegate.window under UISceneDelegate
**Severity:** S1 — Critical (crash on launch)
**Found by:** Pre-Sprint-3 manual validation (Phase 0). iOS Simulator (iPhone 16 / iOS 18.6), Flutter 3.44.2, Xcode 26.5. Reproduces on any iOS target (build-level, not simulator-specific).
**Status:** Fix applied in the working tree, pending commit/PR.

### Repro
Launch the iOS app (Flutter 3.44.2).

### Actual
Fatal crash during `GeneratedPluginRegistrant` plugin registration:
```
flutter_contacts-1.1.9+2 SwiftFlutterContactsPlugin.swift:435:
Fatal error: Unexpectedly found nil while unwrapping an Optional value
```
Line 435: `let rootViewController = UIApplication.shared.delegate!.window!!.rootViewController!`

### Root cause
The iOS project uses the **UISceneDelegate lifecycle** (`UIApplicationSceneManifest` in `Info.plist` + `SceneDelegate.swift` + `FlutterImplicitEngineDelegate`), so `AppDelegate.window` is nil. `flutter_contacts` 1.1.9+2 predates UIScene and force-unwraps it. Fixed upstream in 2.x (2.2.0: "requires Flutter 3.44+").

### Fix (applied)
- `pubspec.yaml`: `flutter_contacts: ^1.1.9+2` -> `^2.2.2`.
- Migrated `lib/features/friends/data/contact_service.dart` to the 2.x API: `permissions.check/request(PermissionType.read)` (replaces removed `requestPermission()`); `getAll(properties: ContactProperties.allProperties)` (replaces `getContacts(withProperties:)`); `displayName ?? ''` (2.x `Contact.displayName` is nullable); added `_mapPermissionStatus` (granted/limited->granted; permanentlyDenied/restricted->deniedPermanently), which also restores the SCR-10 Open-Settings path the old code never reached.
- Regenerated `pubspec.lock` and `ios/Podfile.lock`.
- Verified: `flutter analyze` clean; friends tests pass (incl. S6/PY3 add-friend e2e); app launches on iPhone 16 sim and wires to the emulator.

### Test gap
CI builds iOS but never *launches* it. Recommend an iOS launch smoke test (integration_test) on the CI iOS lane.

**Assigned:** Flutter Dev.

### LINKED ISSUE #96: [S2] Cold-start notification deep-link never runs: ref.read after dispose in NotificationsLifecycleHost
**Severity:** S2 — Major (cold-start notification deep-link broken); graceful, no crash.
**Found by:** Pre-Sprint-3 manual validation (Phase 0; observed on every launch). iOS Simulator.
**Note:** Pre-existing; previously masked by the launch crash #95.

### Repro
Launch the app (cold start).

### Actual
`_NotificationsLifecycleHostState._handleColdStart` (`lib/features/notifications/presentation/notifications_lifecycle_host.dart:92`) calls `ref.read(firebaseMessagingProvider)` after the State element is disposed (scheduled via a warm-up-frame timer; runs post-dispose during splash routing). Riverpod throws `_assertNotDisposed`; caught at line 108 and logged "[NotificationsLifecycleHost] cold start failed". Fires on every launch, so `getInitialMessage()` never runs and the cold-start notification deep-link (FR-AC-05) never executes.

### Expected
On a notification-triggered cold start, the deep link dispatches (or is cached for post-sign-in replay).

### Suggested fix
Guard the timer callback (check disposed/`mounted` before `ref.read`), capture the ref/providers before scheduling, or host cold-start handling in a non-disposed scope (ProviderContainer / top-level provider) rather than widget State.

### Needs
Confirmation in a real notification cold start (Phase 7, production, real device).

**Assigned:** Flutter Dev. Milestone: Sprint 4 (Notifications theme; non-blocking for Sprint 3).

### LINKED ISSUE #97: [S3] OTP cells don't auto-clear after an invalid code (SCR-04): widget controllers not re-synced
**Severity:** S3 — Minor (workaround: manual backspace/retype).
**Found by:** Pre-Sprint-3 manual validation (Phase 1). iOS Simulator (iPhone 16 / iOS 18.6), emulator.

### Repro
On the OTP screen (SCR-04), enter an invalid 6-digit code (e.g. `000000`). Auto-verify fires on the 6th digit.

### Expected (SCR-04 "Error (invalid OTP)")
Cell borders turn `danger`, error text "Incorrect code, try again.", then **cells clear after 1 second** and input re-enables for retry.

### Actual
Error text + red borders appear and retry works (manual backspace/retype), but the cells **do not auto-clear** — the wrong digits remain on screen until manually deleted.

### Likely root cause
`lib/features/auth/application/otp_entry_controller.dart` (Failure case, ~lines 234-240) resets `digits: List.filled(6, '')` in state immediately, but the OTP input widget `lib/features/auth/presentation/widgets/otp_input.dart` does not re-sync its per-cell `TextEditingController`s to the cleared state, so the visual cells retain the old digits. (Also: SCR-04 specifies a 1-second delay before clearing; the controller clears immediately in state.)

### Suggested fix
In `otp_input.dart`, sync `_controllers[i].text` when the controller's `digits` are cleared/changed (e.g. via `didUpdateWidget` or a clear signal). Optionally honour the 1s delay per SCR-04.

**Milestone:** Sprint 5 (Polish). **Assigned:** Flutter Dev.

### LINKED ISSUE #98: [S1] Add-friend hangs forever: duplicate-check get on non-existent friendship denied by rules (Null value error)
**Severity:** S1 — Critical (core add-friend flow blocked end-to-end; affects production).
**Found by:** Pre-Sprint-3 manual validation (Phase 2, S6 re-verification). iOS Simulator, emulator. Reproduces on all targets (rules-level).
**Status:** Client fix applied in working tree, pending commit/PR.

### Repro
Sign in as User A. Friends > Add Friend > Enter Number > a registered user's number > Add Friend.

### Actual
Infinite "looking up" loader; no friendship document is ever written.

### Root cause
1. `lookupUserByPhoneNumber` succeeds server-side (`matched:true`).
2. `MatchAndInviteController.performLookup` (`lib/features/friends/application/match_and_invite_controller.dart:153-158`) then calls `friendshipExists()`, which does `friendships/{sortedId}.get()` on a doc that **does not exist yet**.
3. The friendships read rule (`firestore.rules:113` — `allow read: if request.auth.uid in resource.data.memberIds`) raises a **Null value error** for a non-existent doc (`resource.data` is null) → the `get` is denied with `permission-denied`.
4. `performLookup` does **not** catch it → state stays `MatchAndInviteLoading` → infinite loader.

Verified directly against the Firestore emulator: reading the non-existent `friendships/{sortedId}` as the authenticated member returns `403 ... Null value error ... @ L113`.

### Why this slipped through
`test/features/friends/add_friend_flow_e2e_test.dart` uses a **fake** `FriendshipStore` whose `exists()` returns cleanly — it never exercises the real security rules. The strict read rule is *correct* (it prevents social-graph enumeration), so the fix belongs in the client.

### Fix (applied)
`lib/features/friends/data/friendship_repository.dart` — `FirestoreFriendshipStore.exists` now catches `FirebaseException(permission-denied)` and returns `false` (a member can always read an *existing* friendship, so a denial means the doc doesn't exist yet). Rules untouched.

### Follow-up (recommended)
Add a real-rules regression test (Firestore-rules emulator or a friends integration test) so the duplicate-check-against-rules path is covered, not just the fake store.

**Assigned:** Flutter Dev (fix) / Architect (rules reviewed, intentionally kept strict).

### LINKED ISSUE #99: [S2] Add-friend: no feedback or navigation after 'Add as friend' (friendship created, UI does nothing)
**Severity:** S2 — Major (core add-friend completion has no UI feedback/navigation).
**Found by:** Pre-Sprint-3 manual validation (Phase 2, S6 re-verification). iOS Simulator, emulator. Exposed after fixing the lookup hang #98.

### Repro
Add a friend by manual entry; on the match-confirmation card tap **Add as friend**.

### Actual
The friendship **is** created on the backend (correct shape, no `simplifiedBalances`), but the UI does nothing — it stays on `_MatchFoundCard`. No snackbar, no navigation. The user appears stuck (and may tap repeatedly). Workaround: navigate back -> the friend is in the list.

### Root cause
`MatchAndInviteController.addFriend` (`lib/features/friends/application/match_and_invite_controller.dart:196-216`) creates the friendship and logs analytics but **does not change state on success** (stays `MatchAndInviteMatchFound`). `MatchAndInviteScreen` (`lib/features/friends/presentation/match_and_invite_screen.dart`) has no success state, no `ref.listen`, and no navigation.

### Expected (SCR-10)
On successful add: navigate to Friend Detail (SCR-11) and/or show "[Name] added as a friend", then return to the friends list.

### Suggested fix
Add a success state (e.g. `MatchAndInviteAdded`) emitted by `addFriend`, and have the screen `ref.listen` to navigate to Friend Detail / show the snackbar / pop.

**Assigned:** Flutter Dev.

### LINKED ISSUE #100: [S3] Newly-added friend shows 'Unknown' until relaunch (userProfileProvider caches creation-time null)
**Severity:** S3 — Minor (cosmetic; self-heals on relaunch).
**Found by:** Pre-Sprint-3 manual validation (Phase 2). iOS Simulator, emulator.

### Repro
As User A, add User B as a friend, then view the Friends list **without relaunching**.

### Actual
The just-added friend shows as **"Unknown"** on the creator's side until the app is relaunched (then it correctly shows the name). The friendship data is correct; only the displayed name is wrong.

### Root cause
`userProfileProvider` (`lib/features/friends/application/user_profile_provider.dart`) is a cached `FutureProvider.family<UserModel?, String>`. When the friends list rebuilds at friendship-creation time, it reads the other member's profile (`friends_list_provider.dart:109`) and gets `null` — a race where the users read rule's `isFriendshipMemberWith` `get(friendship)` transiently fails before the new friendship is visible. `friends_list_provider.dart:123` then renders `displayName: profile?.displayName ?? 'Unknown'`. The `null` is cached by the family provider and never refreshed. Rules DO allow the read (verified: User A can read User Two -> "User Two").

### Suggested fix
Invalidate/refresh `userProfileProvider(otherUserId)` after creating a friendship, or make the provider `autoDispose` + retry on `null`, or re-read the profile when the friendship stream emits a new member.

### Related
Exposed alongside #98 (lookup hang). The add-friend completion path also has #<D5> (no navigation).

**Milestone:** Sprint 5 (Polish). **Assigned:** Flutter Dev.

### LINKED ISSUE #101: [S2] Friend Detail broken: settlements watchByContext list query denied by from/to-only read rule
**Severity:** S2 — Major (Friend Detail screen broken end-to-end; blocks balance/history view and Phase 3/4 display validation). Affects production.
**Found by:** Pre-Sprint-3 manual validation (Phase 2). iOS Simulator, emulator. Reproduces on all targets (rules-level).

### Repro
Open Friend Detail (tap a friend row) for any friendship.

### Actual
Screen shows **"Something went wrong" + Retry**. The Friend Detail provider joins three streams; the **settlements** stream errors.

### Root cause
`SettlementRepository.watchByContext` (`lib/features/settlements/data/settlement_repository.dart:90-97`) issues a **list query**: `settlements where contextType == 'friendship' && contextId == <fid>`. The settlements read rule (`firestore.rules:491-493`) only authorizes **per-document** reads:
```
allow read: if request.auth != null
  && (request.auth.uid == resource.data.fromUserId
      || request.auth.uid == resource.data.toUserId);
```
Firestore cannot authorize a `contextId`-filtered list query against a `fromUserId`/`toUserId` rule, so the query is denied. Verified via the emulator REST API:
`403 PERMISSION_DENIED — "Property fromUserId is undefined on object. for 'list' @ L491"`.

`friend_detail_provider` surfaces the stream error as `AsyncError` → the screen renders the error state.

### Why this slipped through
Friend Detail / settlement tests use fake repositories; the real security rules are never exercised by the `watchByContext` list query.

### Suggested fix (architect)
Make the settlements **read** rule authorize the context-member list query, e.g.:
```
allow read: if request.auth != null
  && isContextMember(resource.data.contextType, resource.data.contextId, request.auth.uid);
```
(`isContextMember` already exists at `firestore.rules:418`.) This is semantically correct — any member of the friendship/group context may read its settlements — and matches the `contextType`+`contextId` equality filters of the query. Add a Firestore-rules emulator test for the `watchByContext` list query. Verify create/update rules and the extension-point locks are unchanged (Invariant-safe).

**Assigned:** Architect (rules) + QA (regression test).

### LINKED ISSUE #102: [S1] Add-expense context picker can't load friends: modal shown outside the currentUserId ProviderScope
**Severity:** S1 — Critical (core add-expense flow blocked at step 1; affects production). Deterministic.
**Found by:** Pre-Sprint-3 manual validation (Phase 3). iOS Simulator, emulator. Reproduces on all targets.

### Repro
Tap the **+ FAB** on the authenticated shell to open the add-expense context picker.

### Actual
The picker shows **"Something went wrong loading your friends"** (Retry does not help). No expense can be started via the FAB.

### Root cause
- `AuthenticatedShell._onFabTapped` shows the picker via `showModalBottomSheet(context: context, ...)` (`lib/features/shell/presentation/authenticated_shell.dart:123-128`) with the default `useRootNavigator: false` → the **root** Navigator (the shell has no nested Navigator above it).
- The per-arm `ProviderScope` that overrides `currentUserIdProvider` / `currentUserPhoneProvider` is inside `MaterialApp.home` (`lib/main.dart:176-182`), i.e. **below** the root Navigator.
- The modal route therefore renders **outside** that scope. `AddExpenseContextPickerSheet` watches `friendsListProvider`, which reads `currentUserIdProvider` — and the unscoped provider **throws `UnimplementedError`** (`lib/features/friends/application/friends_list_provider.dart:16`). The `StreamProvider` resolves to `AsyncError` → the picker's error sub-state.

### Why this slipped through
Widget tests for the picker run inside a test `ProviderScope` that overrides `currentUserIdProvider` globally, so the modal-vs-scope mismatch never surfaces.

### Suggested fix
Re-establish the scope at the sheet builder, reading the values from the shell scope (which has them):
```dart
final uid = ref.read(currentUserIdProvider);
final phone = ref.read(currentUserPhoneProvider);
showModalBottomSheet(
  context: context, isScrollControlled: true, useSafeArea: true,
  builder: (_) => ProviderScope(
    overrides: [
      currentUserIdProvider.overrideWithValue(uid),
      currentUserPhoneProvider.overrideWithValue(phone),
    ],
    child: const AddExpenseContextPickerSheet(),
  ),
);
```
Consider auditing all root-Navigator modals/routes for the same scope gap, or hoisting the per-arm scope above `MaterialApp`. Add a test that pumps the sheet via the real root-Navigator path.

**Assigned:** Flutter Dev / Architect (scope architecture).

### LINKED ISSUE #104: [S3] After deleting an expense, screen stays on the stale deleted-expense view instead of returning to Friend Detail
## Summary
After deleting an expense, the screen stays on the now-deleted expense detail view instead of returning to Friend Detail (SCR-11). The user is stranded on a stale screen showing the deleted expense.

## Repro
1. As the expense creator, open an expense detail (SCR-21).
2. Tap Delete -> confirm.
3. Backend delete succeeds and balances recompute, but the screen does not navigate back.

## Root cause
`expense_detail_screen.dart` gated `Navigator.maybePop()` on re-reading the `addExpenseController` state (autoDispose+family). That controller can be GC'd between reads (per its own comment), so the `Success(deleted)` transition is missed and the pop never fires.

## Fix (this session)
`softDelete()` now returns `bool`; the screen navigates on the return value, independent of provider state. Regression test added for the bool contract. Device-verified on the emulator.

Found during pre-Sprint-3 manual validation (re-validating creator-only edit/delete #72). Fixed on branch `fix/sprint-1-2-validation-defects` (commit 23460ef).

### LINKED ISSUE #105: [S2] Adding an expense throws Firebase Analytics assertions (boolean params) and slows the save
## Summary
Adding an expense throws unhandled assertions and makes the save very slow in debug builds. Firebase Analytics only accepts String|num parameter values, but the expense controller emits **boolean** params: `payer_is_self`, `has_receipt`, `has_notes`, `is_offline`.

## Repro
1. Add any expense (debug build).
2. Console shows repeated `Unhandled Exception: 'value is String || value is num' must be set as the value of the parameter: payer_is_self. true found instead` (and `has_receipt`, etc.) from `_emitSaveSucceeded` / `_emitStep2Completed` / `_emitStep3Opened`.
3. The save flow is bogged down by the exception reporting; in release the params would be malformed/dropped.

## Root cause
`FirebaseAnalyticsService.logEvent` forwarded `Map<String, Object>` params raw; the expense controller passed bool values, which `firebase_analytics` asserts against.

## Fix (this session)
Centralised coercion `bool -> "true"/"false"` in `FirebaseAnalyticsService` via a new `sanitiseAnalyticsParameters` helper, so every event is wire-correct. Regression test added. Device-verified.

Found during pre-Sprint-3 manual validation. Fixed on branch `fix/sprint-1-2-validation-defects` (commit 23460ef).

### LINKED ISSUE #106: [S1] Add-friend duplicate-check stalls ~90-120s (denied/one-shot get on iOS-sim+emulator)
## Summary
The add-friend lookup took ~90-120s on the iOS simulator + Firestore emulator. The match card eventually appeared, but the delay made the flow feel broken.

## Root cause
The duplicate-check did a `get()` on the deterministic (non-existent) friendship document. Two problems:
- For a non-existent friendship the rules deny the `get` (Null value error on `resource.data.memberIds`); the mobile SDK does not fail-fast on that denial.
- Against the Firestore emulator on the iOS simulator a one-shot server `get` stalls ~90s (FlutterFire #10449), versus ~2ms for a cached listener read.

Instrumented timing: `friendshipExists` 90043ms (`.get()`) vs 2.2ms (`.snapshots().first`). REST get of the same doc server-side = 7.7ms (instant) — so the stall is entirely client-side.

## Fix (this session)
Replaced `exists(path)` with `existsForMember(memberId, friendshipId)`: a membership query (`memberIds array-contains memberId`) resolved via `.snapshots().first`, reading the cache the always-on friends-list listener keeps fresh. Always allowed for the caller's own friendships; accurate and offline-resilient. Device-verified 90s -> 2.2ms.

Found during pre-Sprint-3 manual validation. Fixed on branch `fix/sprint-1-2-validation-defects` (commit 2cba3cc).

> Note: the underlying iOS-sim + emulator slow-`.get()` behaviour (#10449) is an environment limitation, not a production bug; it also affects other one-shot reads (e.g. profile resolution, the dashboard spend card) on that combo only.

### LINKED ISSUE #107: [S3] Expense detail shows 'Friend' instead of the friend's name in Paid-by/split rows
## Summary
The read-only expense detail (SCR-22) showed "Friend" (and "You") instead of the friend's actual display name in the Paid-by row and the split rows. Surfaced when deep-linking from the activity feed — the user couldn't tell who the expense was with.

## Root cause
`expense_detail_screen.dart` hard-coded `isMyExpense ? 'You' : 'Friend'` (line 275) and `split.userId == currentUserUid ? 'You' : 'Friend'` (line 319).

## Fix (this session)
Resolve the other participant's display name via `userProfileProvider(otherUserUid)` (kept warm by the always-on friends-list listener, so it is an instant cache read in practice); fall back to "Friend" while loading or if unresolved. The current user's rows still read "You". Widget test added. Device-verified.

Found during pre-Sprint-3 manual validation. Fixed on branch `fix/sprint-1-2-validation-defects` (commit c24fbec).

### LINKED ISSUE #108: [S3] Deleted accounts render as 'Unknown' instead of 'Deleted User' (tombstone fails to parse)
## Summary
After a friend deletes their account, a surviving friend sees them as **"Unknown"** instead of the intended **"Deleted User"** label.

## Root cause
The deletion cascade (FR-AU-09) tombstones `users/{uid}` into a PII-free shell `{displayName: "Deleted User", deletedAt}` (ADR-0016), stripping `phoneNumber`, `photoUrl`, `fcmTokens`, etc. But `UserModel.fromFirestore` cast `phoneNumber`/`displayName` with a non-nullable `as String`, so parsing the tombstone **threw**; `friendsListProvider._projectItem` caught the error and fell back to `displayName ?? 'Unknown'`.

## Fix (this session)
`UserModel.fromFirestore` now parses `phoneNumber`/`displayName` as `String? ?? ''`, matching the already-nullable handling of the other stripped fields, so the tombstone parses and "Deleted User" surfaces. Surviving friendships/expenses untouched (Invariant 2). Added a `UserModel.fromFirestore` test (full doc + tombstone). Device-verified.

Found during pre-Sprint-3 manual validation (delete-account cascade, FR-AU-09). Fixed on branch `fix/sprint-1-2-validation-defects` (commit 51c83f4).

### LINKED ISSUE #109: [S3] OTP screen overflows at large Dynamic Type (vertical) and on small phones (horizontal)
## Summary
The OTP entry screen (SCR-04) overflows in two cases: (a) **vertically** at large Dynamic Type, and (b) **horizontally** on small phones (iPhone SE / mini) regardless of text size.

## Root cause
- (a) The screen body was a non-scrolling `Column`; large text pushed the stacked content past the screen height.
- (b) The six cells were fixed at `width: 48`, so 6×48 + 5×12 gaps + 48 padding = **396 dp** exceeds a 320–375 dp phone width.

## Fix (this session)
- Wrapped the OTP body in a `SingleChildScrollView` (vertical).
- `OtpInput` now sizes the cell and gap responsively via `LayoutBuilder` (≤48 dp, shrinking to fit available width) so the row never overflows.
- Added a 320 dp widget regression test (no RenderFlex overflow).

Device-verified the vertical scroll fix; the horizontal fit is proven by the 320 dp test (no iPhone SE simulator installed). Found during pre-Sprint-3 manual validation (Phase 10 accessibility). Fixed on branch `fix/sprint-1-2-validation-defects` (commit 80762d0).

### PR DESCRIPTION
**Pull request by avtanshgupta**
## Summary

Human-in-the-loop **manual validation of everything shipped in Sprint 1 + Sprint 2**, run end-to-end against the **Firebase Emulator Suite** across two real targets (iPhone 16 + iPhone 17 Pro simulators, two distinct signed-in users), as the pre-Sprint-3 readiness gate.

Every scenario was confirmed on **both** the app UI **and** the Firebase side (Firestore, Auth, Storage, Functions logs in the Emulator UI). The pass surfaced **14 defects (D1–D14)**, all fixed, regression-tested, and re-verified in a second clean run.

**Verdict: Pass 1 GREEN** — the shipped Sprint-1 + Sprint-2 surface is validated and free of blocking defects on the emulator.

## Defects fixed

| ID | Sev | Phase | Issue | Summary |
|----|-----|-------|-------|---------|
| D1 | S1 | 0 Build | #95 | iOS simulator build failure blocked launch |
| D2 | S2 | 7 Notif | #96 | Cold-start notification handler crash on null initial message |
| D3 | S3 | 1 Auth | #97 | OTP cells did not clear after a wrong code |
| D4 | S1 | 2 Friends | #98 | Add-friend match branch never created the friendship (S6) |
| D5 | S2 | 2 Friends | #99 | No navigation/confirmation after friendship create |
| D6 | S3 | 2 Friends | #100 | Newly-added friend showed blank display name |
| D7 | S2 | 2 Friends | #101 | Friend Detail context-member read denied by rules |
| D8 | S1 | 3 Expenses | #102 | Context picker empty — could not add an expense |
| D9 | S1 | 2 Friends | #106 | Add-friend duplicate-check stalled ~90s (FlutterFire #10449) |
| D10 | S3 | 3 Expenses | #104 | Post-delete navigation stayed on a stale screen |
| D11 | S2 | 3 Expenses | #105 | Analytics threw on bool params, slowing expense save |
| D12 | S3 | 6 Activity | #107 | Expense detail showed "Friend"/"You", not names |
| D13 | S3 | 8 Profile | #108 | Deleted account rendered as "Unknown", not "Deleted User" |
| D14 | S3 | 10 X-cut | #109 | OTP screen overflowed at large text and on small phones |

Severity via the `triage-bug` skill. None remain blocking.

## Commits

| Commit | Scope |
|--------|-------|
| `0f891ba` | D1–D8 (#95–102) |
| `2cba3cc` | D9 add-friend stall (#106) |
| `23460ef` | D10 nav + D11 analytics bool (#104, #105) |
| `c24fbec` | D12 expense-detail names (#107) |
| `51c83f4` | D13 deleted-user label (#108) |
| `80762d0` | D14 OTP overflow (#109) |
| `3651e69` | Final QA report |

## Deliverable

`docs/qa/manual-validation-sprint-1-2.md` — per-scenario pass/fail matrix for Phases 0–10, defect list with issue links, FlutterFire #10449 environment notes, and the Pass 1 sign-off.

## Invariants

All four spot-checked and explicitly validated in Phase 9: INV1 integer paise (exact split sums, no float drift), INV2 `simplifiedBalances` server-only (direct client write rejected by rules), INV3 system share sheet, INV4 single Firebase project. Security rules verified: non-member / stranger / unauthenticated / enumeration all denied.

## Verification

- `fvm flutter test` — full suite green (1609 tests)
- `fvm flutter analyze lib test` — no issues
- `dart format .` — clean

## Scope note

This PR is **Pass 1 (emulator) only**. Pass 2 (production + real device) is **deferred** — its findings (D15 backend deploy, D16 iOS push capability, D17/D18 App Check) are tracked separately and are **not** part of this PR. No production-only or Pass-2 code changes are included here.

Closes #95
Closes #96
Closes #97
Closes #98
Closes #99
Closes #100
Closes #101
Closes #102
Closes #104
Closes #105
Closes #106
Closes #107
Closes #108
Closes #109

Co-authored-by: Copilot <223556219+Copilot@users.noreply.github.com>

### CODE REVIEW

File: firestore.rules [hunk 1]
```diff
@@ -487,10 +487,18 @@ service cloud.firestore {
...
+      // SCR-11 Friend Detail). Per-doc and list reads both resolve through
+      // the contextType/contextId equality filters.
       allow read: if request.auth != null
-        && (request.auth.uid == resource.data.fromUserId
-            || request.auth.uid == resource.data.toUserId);
+        && isContextMember(
+          resource.data.contextType,
+          resource.data.contextId,
+          request.auth.uid
+        );
```
Comment (avtanshgupta) [inline firestore.rules:501]: **[Major] Security-rule change without a rules test for the path it fixes (D7).**

**Problem:** The read rule moved from `uid == fromUserId || uid == toUserId` to `isContextMember(...)` (which `get()`s the parent friendship/group doc). The D7 bug was the `contextId`-filtered **list** query (`SettlementRepository.watchByContext`, `lib/features/settlements/data/settlement_repository.dart:90-96`), but the settlement read tests (`functions/test/firestore-rules/settlements.test.ts:204-237`) are all single-doc `getDoc` reads, and this PR doesn't add one.

**Impact:** For a friendship `isContextMember` is equivalent to from/to, so the existing single-doc tests pass unchanged — a green rules job does **not** lock in the D7 fix, and a future regression of the list-read path would ship silently.

**Fix:** Add two rules tests — a context member listing settlements by `contextId` (`getDocs(query(coll, where('contextId','==',FID)))`) succeeds, and a non-member's identical list query is denied.

File: lib/features/auth/domain/user_model.dart [hunk 2]
```diff
@@ -25,8 +25,13 @@ class UserModel {
...
     return UserModel(
-      phoneNumber: data['phoneNumber'] as String,
-      displayName: data['displayName'] as String,
+      // FR-AU-09 (D13): the account-deletion tombstone replaces the doc
+      // with a PII-free shell ({displayName: "Deleted User", deletedAt}),
+      // so phoneNumber is absent. Tolerate it (default to '') instead of
+      // a hard cast that throws and makes the friend render as "Unknown"
+      // rather than the intended "Deleted User".
+      phoneNumber: data['phoneNumber'] as String? ?? '',
+      displayName: data['displayName'] as String? ?? '',
```
Comment (avtanshgupta) [inline lib/features/auth/domain/user_model.dart]: **[Minor] `displayName` null-default can mask malformed user docs.**

**Problem:** The FR-AU-09 tombstone always sets `displayName: 'Deleted User'`, so only `phoneNumber` is actually absent — `displayName` never needs the `?? ''` fallback for D13.

**Impact:** A genuinely malformed real user doc (missing `displayName`) would now render as a blank name instead of failing loudly.

**Fix:** Keep the null-tolerant cast on `phoneNumber` (the real tombstone case); consider a hard cast or an explicit log when `displayName` is unexpectedly absent.

File: lib/features/auth/presentation/widgets/otp_input.dart [hunk 3]
```diff
@@ -138,37 +167,64 @@ class _OtpInputState extends State<OtpInput> {
...
+          return Row(
+            mainAxisAlignment: MainAxisAlignment.center,
+            children: List.generate(_cellCount, (index) {
+              return Padding(
+                padding: EdgeInsets.only(
+                  right: index < _cellCount - 1 ? gap : 0,
+                ),
+                child: SizedBox(
+                  width: cell,
+                  height: maxCell,
```
Comment (avtanshgupta) [inline lib/features/auth/presentation/widgets/otp_input.dart:200]: **[Minor] OTP cell width drops below the 48 dp tap target on narrow screens.**

**Problem:** With the screen's 24 dp horizontal padding on a 320 dp viewport, the six cells compute to ~37.5 dp wide (`height` stays 48 dp).

**Impact:** Below the 48x48 dp WCAG / accessibility-spec target on small phones (e.g. iPhone SE).

**Fix:** Acceptable trade-off vs. the D14 overflow (6x48 + gaps can't fit 320 dp), and better than overflowing — just flagging the residual a11y gap. The new 320 dp regression test is a good guard.

File: lib/features/friends/presentation/match_and_invite_screen.dart [hunk 4]
```diff
@@ -14,6 +15,25 @@ class MatchAndInviteScreen extends ConsumerWidget {
     final state = ref.watch(matchAndInviteControllerProvider);
     final controller = ref.read(matchAndInviteControllerProvider.notifier);
 
+    ref.listen<MatchAndInviteState>(matchAndInviteControllerProvider, (
```
Comment (avtanshgupta) [inline lib/features/friends/presentation/match_and_invite_screen.dart:18]: **[Minor] No widget test for the D5/D6 navigation + cache-invalidation wiring.**

**Problem:** The `MatchAndInviteAdded` -> `ref.invalidate(userProfileProvider)` + `Navigator.pop` + snackbar path is only covered indirectly via the controller-state test (`match_and_invite_controller_test.dart` emits `Added`, but not the screen reaction).

**Impact:** A regression in the screen wiring (the `ref.listen` condition, the pop, or the invalidate) would not be caught.

**Fix:** Add a widget test asserting the snackbar text and that the route pops on `MatchAndInviteAdded`.

**Review by avtanshgupta — COMMENTED** [review]
## Review — Sprint 1-2 Pass 1 emulator defects (D1-D14)

Reviewed the full diff against the four invariants, the changed Firestore security rule, Riverpod/UI/async handling, telemetry/PII, CI gates, test surfaces, and milestone reconciliation.

**Verdict:** No blocking issues. All four invariants hold (no float money, no client `simplifiedBalances` write, no app-specific share target, single Firebase project untouched). High-quality defect-fix batch with regression tests for most defects. One security-test gap is worth addressing before merge (inline on `firestore.rules`).

### Major
- `firestore.rules` — the settlements read-rule change (D7) ships without a rules test for the `contextId` list-query path it actually fixes. See inline.

### Recommendations (Minor)
- `user_model.dart` — `displayName` null-default can mask malformed docs (inline).
- `otp_input.dart` — OTP cell width drops below 48 dp on 320 dp screens (inline).
- `match_and_invite_screen.dart` — no widget test for the D5/D6 nav/invalidate wiring (inline).
- PR body uses a narrative format rather than the template's Invariant Checklist + per-scope coverage lines; consider milestoning this PR to **Sprint 2** (the 14 closed issues are all correctly on Sprint 2; merging leaves 2 open, so the milestone stays open — no close needed).

### Approved
- **D8** root-Navigator scope move (`lib/main.dart`) — correct and well-tested (`root_navigator_scope_test.dart`); verified `NotificationsLifecycleHost` reads only root providers, so wrapping it outside the scope is safe.
- **D9** `existsForMember` membership query — rules-safe emulator workaround; fakes updated consistently. The `as String` cast is legitimate (`_friendshipRepository` is `dynamic`, `createFriendship` returns `Future<String>`).
- **D10–D14** — `softDelete` bool return, analytics bool coercion, expense-detail name fallback (`.maybeWhen`), tombstone tolerance, and OTP responsiveness all carry regression tests.
- flutter_contacts 1.x to 2.x migration — exhaustive permission mapping; `pubspec.lock` / `ios/Podfile.lock` consistent.

CI: Flutter Lint & Test, Integration Tests (incl. the Firestore-rules step), Coverage Gate, Build Android, PR Title Lint all pass; Build iOS pending.

File: firestore.rules (hunk 1 above)
Comment (avtanshgupta) [reply firestore.rules:501]: Done in 254c928. Added a `settlements/{sid} — contextId list-read rules (D7)` describe block in `functions/test/firestore-rules/settlements.test.ts` that mirrors `SettlementRepository.watchByContext`: a context member's `where(contextType ==) + where(contextId ==) + orderBy(date)` list query **succeeds** (and returns the two seeded docs), while a non-member's and an unauthenticated identical query are **denied**.

One nuance worth recording: both equality filters are required. The read rule references `resource.data.contextType` **and** `resource.data.contextId`, so a `contextId`-only list query can't resolve the rule and is denied even for a member — the test reproduces the production query exactly (both filters), which is what locks in the list-read path.

Verified locally against the Firestore emulator: 47/47 settlement rules tests pass.

File: lib/features/auth/domain/user_model.dart (hunk 2 above)
Comment (avtanshgupta) [reply lib/features/auth/domain/user_model.dart]: Done in 254c928. Restored `displayName` to the non-nullable `as String` cast and kept only `phoneNumber` null-tolerant (the real FR-AU-09 tombstone case). A user doc missing `displayName` now fails loudly rather than rendering a blank name.

The tombstone always writes `displayName: 'Deleted User'` (and `toCreateMap` always sets it on creation), so D13 is unaffected — `user_model_test.dart` plus the full auth + friends suites (496 tests) stay green.

File: lib/features/auth/presentation/widgets/otp_input.dart (hunk 3 above)
Comment (avtanshgupta) [reply lib/features/auth/presentation/widgets/otp_input.dart:200]: Acknowledged — leaving the layout as-is per your note. Six 48 dp cells plus gaps cannot fit a 320 dp viewport, so shrinking the cell width (height stays 48 dp) is the lesser evil versus the D14 overflow, and the new 320 dp regression test guards it.

Fully closing the residual a11y gap would need a different layout (e.g. wrapping the cells onto two rows, or scaling the whole row), which is out of scope for this Pass-1 PR — flagging it as a follow-up rather than changing the responsive sizing here.

File: lib/features/friends/presentation/match_and_invite_screen.dart (hunk 4 above)
Comment (avtanshgupta) [reply lib/features/friends/presentation/match_and_invite_screen.dart:18]: Done in 254c928. Added a widget test in `test/features/friends/match_and_invite_screen_test.dart` that drives `MatchAndInviteMatchFound -> MatchAndInviteAdded` and asserts the full screen reaction:

- the route **pops** back (the screen's `Add Friend` AppBar is gone and Home is shown again);
- the confirmation **snackbar** reads `'<name> added as a friend'`;
- `userProfileProvider(otherUserId)` is **invalidated** — a persistent Home consumer keeps the family instance alive so the invalidate forces an observable refetch, which the test counts.

So the `ref.listen` condition, the `Navigator.pop`, and the `ref.invalidate` are all now covered directly rather than only via the controller-state test.

**Review by avtansh-code — APPROVED** [review]
